atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
+                            transactionService,
+                            NoOpCleaner.INSTANCE,
+                            () -> startTimestamp,
+                            getConflictDetectionManager(),
+                            SweepStrategyManagers.createDefault(keyValueService),
+                            startTimestamp,
+                            Optional.empty(),
+                            PreCommitConditions.NO_OP,
+                            AtlasDbConstraintCheckingMode.NO_CONSTRAINT_CHECKING,
+                            null,
+                            TransactionReadSentinelBehavior.THROW_EXCEPTION,
+                            false,
+                            timestampValidationReadCache,
+                            getRangesExecutor,
+                            defaultGetRangesConcurrency,
+                            sweepQueueWriter,
+                            deleteExecutor,
+                            validateLocksOnReads,
+                            () -> TRANSACTION_CONFIG),
+                    counter);
+        }
+
+        @Override
+        protected Transaction createTransaction(
+                long immutableTimestamp,
+                Supplier<Long> startTimestampSupplier,
+                LockToken immutableTsLock,
+                PreCommitCondition preCommitCondition) {
+            AtomicInteger counter = new AtomicInteger();
+            return transactionWrapper.apply(
+                    new SerializableTransaction(
+                            metricsManager,
+                            keyValueService,
+                            timelockService,
+                            transactionService,
+                            cleaner,
+                            startTimestampSupplier,
+                            getConflictDetectionManager(),
+                            sweepStrategyManager,
+                            immutableTimestamp,
+                            Optional.of(immutableTsLock),
+                            preCommitCondition,
+                            constraintModeSupplier.get(),
+                            cleaner.getTransactionReadTimeoutMillis(),
+                            TransactionReadSentinelBehavior.THROW_EXCEPTION,
+                            allowHiddenTableAccess,
+                            timestampValidationReadCache,
+                            getRangesExecutor,
+                            defaultGetRangesConcurrency,
+                            sweepQueueWriter,
+                            deleteExecutor,
+                            validateLocksOnReads,
+                            transactionConfig),
+                    counter);
+        }
+    }
+
+    private static class VerifyingDelegate implements AutoDelegate_KeyValueService {
+        private final KeyValueService delegate;
+        private final AtomicInteger atomicInteger;
+
+        VerifyingDelegate(KeyValueService keyValueService, AtomicInteger atomicInteger) {
```

Comment:
I think we should extract a class that wraps the `AtomicInteger`, and the methods are self describing,

e.g.

```java
interface SynchronousGetTracker {
  void checkInAsyncCall();
  void checkInSyncCall();
}
```

something along those lines, having the `counter` spread over a 1000 lines is not great

Reply:
Switched to an interface.